Module set init script

AloisMahdal opened the pull request:
This PR contains two teps:

 *  Implement "module set init" feature, where a module set may contain `init` executable,
    and if it does, p-a will run it before module set and fail if errors are found.

 *  Move current code of `prep_links()` (and associated patch that modifies it) from
     common.sh to module set; hence the other PR in preupgrade-assistant-modules.

AloisMahdal commented:
Note that this PR fixes #280 and obsoletes #279.

AloisMahdal commented:
Thanks for feedback.  The new error handling is as follows:

 *  if file does not exist, nothing happens.

 *  if it does, preupg will always attempt to launch it but any `OSError` is caught and
    re-raised as `ModuleSetFormatError` (a new exception class I've added for purposes
    of malformed module sets).

This should cover more cases than just file is not being executable; also cases when file is a directory or (I think) malformed shebang line.  And yes, preupg crashes; let's err on the safe side.

I also made few minor cleanups within the code added by first commit:

 *  I removed the `return True` at the end; the value is not consumed by anything. (Frankly
     I don't know why I even added it in the first place).

 *  I fixed typo in the `ModuleSetInitError` docstring (`s/file/script/`).

 *  I fixed formatting of `scenario` assignment (spaces around `=` were missing).

AloisMahdal commented:
Oops, made some errors in the last commit, please wait for update.

AloisMahdal commented:
OK, so errors have been fixed:

 *  fixed exception reference (`exception.ModuleSetFormatError`, not just
     `ModuleSetFormatError`)

 *  Improved `ModuleSetFormatError` message to include path and string from OSError.

AloisMahdal commented:
I have also tested it against following cases:

 *  no init,
 *  init is a non-executable file,
 *  init is a directory,
 *  init is executable but empty,
 *  init is executable but with none or bad shebang,
 *  init is executable and passed,
 *  init is executable and failed (stderr or exit code 1).

Also I've re-run result-files test from downstream suite.

pirat89 commented:
It works correctly know. Just in case that the `init` is directory, the error message looks little weird and something like info that the script is expected instead of directory would be better, but it is just littleness and in addition it is improbable situation.

The modules are already prepared for this change so it can be merged.

AloisMahdal commented:
Thanks, updated:

 *  `init_env` now holds just the extra variables. `os.environ` is copied and updated when
    passed to `Popen ` so that we can keep extra variables in `init_env`

 *  `init_env` forms most of the body of the debug message, which is displayed only if
     init is about to be launched.

 *  if init script is not found, nothing is logged.

AloisMahdal commented:
Sorry, I did not realize `dict.update()` updates *in-place* and returns error.

I have pushed new version with fixed dict logic.

AloisMahdal commented:
qa_verified: already done earlier; adding labels in hindsight.
